base/notify.go

```diff
@@ -0,0 +1,247 @@
+
+	// SoftFailure indicates the task and experiment should not fail if the task
+	// cannot successfully send a request to the notification hook
+	SoftFailure bool `json:"softFailure" yaml:"softFailure"`
+}
+
+const (
+	// NotifyTaskName is the task name
+	NotifyTaskName = "notify"
+)
+
+// notifyTask sends notifications
+type notifyTask struct {
+	TaskMeta
+	With notifyInputs `json:"with" yaml:"with"`
+}
+
+// Report is the data that is given to the payload template
+type Report struct {
+	// Group string `json:"group" yaml:"group"`
+
+	// SLOs map[string]SLOReport `json:"SLOs" yaml:"SLOs"`
+
+	// Metrics map[string]MetricReport `json:"metrics" yaml:"metrics"`
+
+	// Timestamp is when the report was created
+	// For example: 2022-08-09 15:10:36.569745 -0400 EDT m=+12.599643189
+	TimeStamp string `json:"timeStamp" yaml:"timeStamp"`
+
+	// Completed is whether or not the experiment has completed
+	Completed bool `json:"completed" yaml:"completed"`
+
+	// NoTaskFailures is whether or not the experiment had any tasks that failed
+	NoTaskFailures bool `json:"noTaskFailures" yaml:"noTaskFailures"`
+
+	// NumTasks is the number of tasks in the experiment
+	NumTasks int `json:"numTasks" yaml:"numTasks"`
+
+	// NumCompletedTasks is the number of completed tasks in the experiment
+	NumCompletedTasks int `json:"numCompletedTasks" yaml:"numCompletedTasks"`
+
+	// NumLoops is the current loop of the experiment
+	NumLoops int `json:"numLoops" yaml:"numLoops"`
+
+	// Experiment is the experiment struct
+	Experiment *Experiment `json:"experiment" yaml:"experiment"`
+}
+
+// NotifyPayloadTemplateValues contains the report as well as the report in some other formats
+type NotifyPayloadTemplateValues struct {
+	// Report is the report
+	Report Report `json:"report" yaml:"report"`
+
+	// JSONStringReport is the report but marshalled into JSON and stringified
+	JSONStringReport string `json:"JSONStringReport" yaml:"JSONStringReport"`
+
+	// EscapedJSONStringReport is the report but marshalled into JSON and stringified and with tabs and new lines escaped
+	EscapedJSONStringReport string `json:"escapedJSONStringReport" yaml:"escapedJSONStringReport"`
+}
+
+// getPayloadTemplateValues gets the values for the payload tempalte
+func getPayloadTemplateValues(exp *Experiment) (*NotifyPayloadTemplateValues, error) {
+	report := Report{
+		// Group: exp.driver.Group
+		// SLOs:           slos,
+		// Metrics:        metrics,
+
+		TimeStamp:         time.Now().String(),
+		Completed:         exp.Completed(),
+		NoTaskFailures:    exp.NoFailure(),
+		NumTasks:          len(exp.Spec),
+		NumCompletedTasks: exp.Result.NumCompletedTasks,
+		NumLoops:          exp.Result.NumLoops,
+		Experiment:        exp,
+	}
+
+	marshalledReport, err := json.MarshalIndent(report, "", "  ")
+	if err != nil {
+		log.Logger.Error("could not JSON marshall report")
+		return nil, err
+	}
+
+	stringReport := string(marshalledReport)
+
+	// escape double quotes, tabs, and new lines
```

Comment:
We really don't want anything special here ... but if special functionality is indeed required, `getTextTemplateFromURL` can be modified so that additional template functions (e.g., for escaping) is introduced as part of the template func map.

Reply:
The idea is to get rid of NotifyPayloadTemplateValues struct altogether and work only with Report.